samza-core/src/main/java/org/apache/samza/coordinator/StreamPartitionCountMonitorFactory.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.coordinator;
+
+import java.util.Set;
+import org.apache.samza.SamzaException;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.JobConfig;
+import org.apache.samza.config.TaskConfig;
+import org.apache.samza.metrics.MetricsRegistry;
+import org.apache.samza.system.StreamMetadataCache;
+import org.apache.samza.system.SystemStream;
+
+
+/**
+ * Common place for reading configs and wiring {@link StreamPartitionCountMonitor}.
+ */
+public class StreamPartitionCountMonitorFactory {
+  private final StreamMetadataCache streamMetadataCache;
+  private final MetricsRegistry metrics;
+
+  public StreamPartitionCountMonitorFactory(StreamMetadataCache streamMetadataCache, MetricsRegistry metrics) {
+    this.streamMetadataCache = streamMetadataCache;
+    this.metrics = metrics;
+  }
+
+  /**
+   * Build {@link StreamPartitionCountMonitor} for input streams.
+   */
+  public StreamPartitionCountMonitor buildInputStreamPartitionCountMonitor(Config config,
```

Comment:
The only reason being is when read `streamPartitionCountMonitorFactory.buildInputStramPartitionCountMonitor(..)` sounds redundant.

I don't mind verbosity but felt it was redundant. I'd prefer conciseness if we can achieve the same intent.

Fair enough, it doesn't have to know about it here. But the regex factory apparently does care about the fact whether its rewritten config vs original config.

Reply:
Ok, I do prefer the conciseness as well, so I will change it back. If we do need differentiation in the future, we can just change the method name.

I suppose it won't hurt to use the name `finalConfig` to help be a little more explicit, but I still don't think the stage of the config should be part of the contract between classes (unless they the classes are directly dealing with config rewriting), because that introduces an ordering dependency. Most classes should be written in a way that is agnostic to where they are being used.

Reply:
I looked more closely at what `ClusterBasedJobCoordinator` was using before. It was using the config from before calculating the job model. It seems like it should ideally be using the config from after calculating job model, but I would prefer to not change current logic in this PR, so that this PR can purely stay as a refactoring PR. I will leave the variable name as `config`, since other variable names like `finalConfig` could be inconsistent with the current logic. I will also create a ticket around this more general problem of having too many "config stages"; most objects should only receive the final config.